components/batch-system/src/batch.rs

```diff
@@ -251,50 +242,46 @@ enum ReschedulePolicy {
 }
 
 impl<N: Fsm, C: Fsm, Handler: PollHandler<N, C>> Poller<N, C, Handler> {
-    fn fetch_batch(&mut self, batch: &mut Batch<N, C>, max_size: usize) {
-        let curr_batch_len = batch.len();
-        if batch.control.is_some() || curr_batch_len >= max_size {
-            // Do nothing if there's a pending control fsm or the batch is already full.
-            return;
+    fn fetch_batch(&mut self, batch: &mut Batch<N, C>) -> bool {
+        if batch.control.is_some() {
+            return true;
         }
 
-        let mut pushed = if curr_batch_len == 0 {
-            match self.fsm_receiver.try_recv().or_else(|_| {
-                self.handler.pause();
-                // Block if the batch is empty.
-                self.fsm_receiver.recv()
-            }) {
-                Ok(fsm) => batch.push(fsm),
-                Err(_) => return,
-            }
-        } else {
-            true
-        };
-
-        while pushed {
-            if batch.len() < max_size {
-                let fsm = match self.fsm_receiver.try_recv() {
-                    Ok(fsm) => fsm,
-                    Err(TryRecvError::Empty) => return,
-                    Err(TryRecvError::Disconnected) => unreachable!(),
-                };
-                pushed = batch.push(fsm);
-            } else {
-                return;
+        if let Ok(fsm) = self.fsm_receiver.try_recv() {
+            return batch.push(fsm);
+        }
+
+        if batch.is_empty() {
+            self.handler.pause();
+            if let Ok(fsm) = self.fsm_receiver.recv() {
+                return batch.push(fsm);
             }
         }
-        batch.clear();
+        true
+    }
+
+    fn try_fetch_batch(&mut self, batch: &mut Batch<N, C>) -> bool {
+        match self.fsm_receiver.try_recv() {
+            Ok(fsm) => batch.push(fsm),
+            Err(_) => true,
+        }
     }
 
     // Poll for readiness and forward to handler. Remove stale peer if necessary.
     fn poll(&mut self) {
         let mut batch = Batch::with_capacity(self.max_batch_size);
         let mut reschedule_fsms = Vec::with_capacity(self.max_batch_size);
 
-        self.fetch_batch(&mut batch, self.max_batch_size);
-        while !batch.is_empty() {
-            let mut hot_fsm_count = 0;
-            self.handler.begin(batch.len());
+        // Fetch batch after every round is finished. It's helpful to protect regions
+        // from becoming hungry if some regions are hot points.
+        let mut run = true;
+        while run && self.fetch_batch(&mut batch) {
+            // If there is some region wait to be deal, we must deal with it even if it has overhead
+            // max size of batch. It's helpful to protect regions from becoming hungry
+            // if some regions are hot points.
+            let max_batch_size = std::cmp::max(self.max_batch_size, batch.normals.len());
+            self.handler.begin(max_batch_size);
```

Comment:
Should always pass `self.max_batch_size`, otherwise there will be probably two allocations.

Reply:
No. Every poll I will try to receive one new Fsm even if the size of batch has reached `self.max_batch_size`. This is to avoid one Fsm is hungry because there are too many Fsms waiting to deal. So the size of batch may exceed `self.max_batch_size`.